* Rationale
- recount cellranger output bam files using different gtf files
- in cellranger this feature is missing and the alignment needs to be redone when recount is wanted
- should only need this folder, the bam file and a gtf file
  - obviously this gtf file needs to be compatible to the aligned genome (version)
* Development
* How to use
* Todo
** TODO handle multiple occuring IDs
Here, it is important to provide an easy way to allow users to specify which ID should be used for counting.
There might be a problem, if features overlap (i.e. exons) so each UMI might be counted multiple times.
Maybe using the geneID in any case might be the best solution and only accept unique UMI - cell barcodes pairs.
